File: ios/scripts/benchmark_webrtc_models.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import csv
import math
from collections import defaultdict
from pathlib import Path

import coremltools as ct
import numpy as np
import soundfile as sf

np.seterr(all="ignore")


ANALYSIS_WINDOW_SAMPLES = 48_000
SPEAKER_THRESHOLD = 0.90
SPOOF_THRESHOLD = 0.665
TARGET_SPOOF_RMS = 0.03
MAX_SPOOF_GAIN = 10.0


def read_csv_rows(path: Path) -> list[dict[str, str]]:
    with path.open(newline="") as fh:
        return list(csv.DictReader(fh))


def load_wav_16k_mono(path: str, target_samples: int = ANALYSIS_WINDOW_SAMPLES) -> np.ndarray:
    samples, sr = sf.read(path, dtype="float32", always_2d=False)
    if samples.ndim > 1:
        samples = samples.mean(axis=1)
    samples = np.nan_to_num(samples, nan=0.0, posinf=0.0, neginf=0.0)
    samples = np.clip(samples, -1.0, 1.0)
    if sr != 16_000:
        raise RuntimeError(f"Unexpected sample rate {sr} for {path}")
    if samples.shape[0] > target_samples:
        samples = samples[-target_samples:]
    elif samples.shape[0] < target_samples:
        samples = np.pad(samples, (0, target_samples - samples.shape[0]))
    return samples.astype(np.float32)


def build_mel_filterbank(num_filters: int = 80, fft_size: int = 400, sample_rate: float = 16_000.0) -> np.ndarray:
    half = fft_size // 2 + 1

    def hz_to_mel(hz: float) -> float:
        return 2595.0 * math.log10(1.0 + hz / 700.0)

    def mel_to_hz(mel: float) -> float:
        return 700.0 * (10.0 ** (mel / 2595.0) - 1.0)

    mel_min = hz_to_mel(0.0)
    mel_max = hz_to_mel(sample_rate / 2.0)
    mel_points = [
        mel_to_hz(mel_min + i * (mel_max - mel_min) / (num_filters + 1))
        for i in range(num_filters + 2)
    ]
    bins = [int((hz / sample_rate) * fft_size + 0.5) for hz in mel_points]

    fb = np.zeros((num_filters, half), dtype=np.float32)
    for m in range(num_filters):
        left, center, right = bins[m], bins[m + 1], bins[m + 2]
        if center > left:
            for k in range(left, min(center, half)):
                fb[m, k] = (k - left) / (center - left)
        if right > center:
            for k in range(center, min(right, half)):
                fb[m, k] = (right - k) / (right - center)
    return fb


MEL_FILTERS = build_mel_filterbank()
HAMMING = np.hamming(400).astype(np.float32)


def compute_fbank(samples: np.ndarray) -> np.ndarray:
    x = samples.astype(np.float32)
    x = np.nan_to_num(x, nan=0.0, posinf=0.0, neginf=0.0)
    x = np.clip(x, -1.0, 1.0)
    emphasized = np.empty_like(x)
    emphasized[0] = x[0]
    emphasized[1:] = x[1:] - 0.97 * x[:-1]

    frames: list[np.ndarray] = []
    for frame_idx in range(300):
        start = frame_idx * 160
        chunk = emphasized[start : start + 400]
        if chunk.shape[0] < 400:
            chunk = np.pad(chunk, (0, 400 - chunk.shape[0]))
        spec = np.fft.rfft(chunk * HAMMING)
        power = (np.real(spec) ** 2 + np.imag(spec) ** 2).astype(np.float32)
        power = np.nan_to_num(power, nan=0.0, posinf=1e10, neginf=0.0)
        energy = MEL_FILTERS @ power
        energy = np.maximum(energy, 1e-10)
        frames.append(np.log(energy))

    fbank = np.stack(frames, axis=0).astype(np.float32)
    mean = float(np.mean(fbank))
    return fbank - mean


class SpeakerBench:
    def __init__(self, model_path: Path):
        self.model = ct.models.MLModel(str(model_path), compute_units=ct.ComputeUnit.CPU_ONLY)

    def embed(self, samples: np.ndarray) -> np.ndarray:
        fbank = compute_fbank(samples)
        out = self.model.predict({"fbank_features": fbank[np.newaxis, :, :]})
        emb = np.array(out["embedding"], dtype=np.float32).reshape(-1)
        return emb

    @staticmethod
    def cosine(a: np.ndarray, b: np.ndarray) -> float:
        denom = float(np.linalg.norm(a) * np.linalg.norm(b))
        if denom <= 1e-8:
            return 0.0
        return max(0.0, min(1.0, float(np.dot(a, b) / denom)))


class SpoofBench:
    def __init__(self, model_path: Path):
        self.model = ct.models.MLModel(str(model_path), compute_units=ct.ComputeUnit.CPU_ONLY)

    def predict_clone_probability(self, samples: np.ndarray) -> float:
        rms = float(np.sqrt(np.mean(samples * samples)))
        gain = min(MAX_SPOOF_GAIN, TARGET_SPOOF_RMS / max(rms, 1e-6))
        prepared = np.clip(samples * gain, -1.0, 1.0).astype(np.float32)
        out = self.model.predict({"waveform": prepared[np.newaxis, :]})
        val = out["clone_probability"]
        if isinstance(val, np.ndarray):
            return float(val.reshape(-1)[0])
        return float(val)


def run_speaker_benchmark(manifest_path: Path, model_path: Path, max_speakers: int) -> dict[str, float]:
    rows = read_csv_rows(manifest_path)
    rows = [r for r in rows if r["label"] == "real" and r["split"] in {"val", "test"}]
    by_speaker: dict[str, list[dict[str, str]]] = defaultdict(list)
    for row in rows:
        by_speaker[row["speaker_id"]].append(row)

    selected = [speaker for speaker, items in sorted(by_speaker.items()) if len(items) >= 2][:max_speakers]
    bench = SpeakerBench(model_path)

    enroll_embeddings: dict[str, np.ndarray] = {}
    pos_scores: list[float] = []
    neg_scores: list[float] = []

    for speaker in selected:
        print(f"speaker_progress={speaker}", flush=True)
        enroll = load_wav_16k_mono(by_speaker[speaker][0]["audio_path"])
        probe = load_wav_16k_mono(by_speaker[speaker][1]["audio_path"])
        enroll_emb = bench.embed(enroll)
        probe_emb = bench.embed(probe)
        enroll_embeddings[speaker] = enroll_emb
        pos_scores.append(bench.cosine(enroll_emb, probe_emb))

    for idx, speaker in enumerate(selected):
        other = selected[(idx + 1) % len(selected)]
        if other == speaker:
            continue
        probe = load_wav_16k_mono(by_speaker[other][0]["audio_path"])
        probe_emb = bench.embed(probe)
        neg_scores.append(bench.cosine(enroll_embeddings[speaker], probe_emb))

    total = len(pos_scores) + len(neg_scores)
    correct = sum(score >= SPEAKER_THRESHOLD for score in pos_scores) + sum(
        score < SPEAKER_THRESHOLD for score in neg_scores
    )
    return {
        "selected_speakers": float(len(selected)),
        "positive_pairs": float(len(pos_scores)),
        "negative_pairs": float(len(neg_scores)),
        "speaker_accuracy": correct / max(total, 1),
        "positive_mean_similarity": float(np.mean(pos_scores)) if pos_scores else 0.0,
        "negative_mean_similarity": float(np.mean(neg_scores)) if neg_scores else 0.0,
    }


def run_spoof_benchmark(manifest_path: Path, model_path: Path, max_per_class: int) -> dict[str, float]:
    rows = read_csv_rows(manifest_path)
    rows = [r for r in rows if r["split"] in {"val", "test"}]
    real_rows = [r for r in rows if r["label"] == "real"][:max_per_class]
    fake_rows = [r for r in rows if r["label"] == "fake"][:max_per_class]
    bench = SpoofBench(model_path)

    real_scores: list[float] = []
    fake_scores: list[float] = []

    for row in real_rows:
        print(f"spoof_progress=real:{Path(row['audio_path']).name}", flush=True)
        real_scores.append(bench.predict_clone_probability(load_wav_16k_mono(row["audio_path"])))
    for row in fake_rows:
        print(f"spoof_progress=fake:{Path(row['audio_path']).name}", flush=True)
        fake_scores.append(bench.predict_clone_probability(load_wav_16k_mono(row["audio_path"])))

    total = len(real_scores) + len(fake_scores)
    correct = sum(score < SPOOF_THRESHOLD for score in real_scores) + sum(
        score >= SPOOF_THRESHOLD for score in fake_scores
    )
    return {
        "real_rows": float(len(real_scores)),
        "fake_rows": float(len(fake_scores)),
        "spoof_accuracy": correct / max(total, 1),
        "real_mean_clone_probability": float(np.mean(real_scores)) if real_scores else 0.0,
        "fake_mean_clone_probability": float(np.mean(fake_scores)) if fake_scores else 0.0,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--speaker-manifest", type=Path, default=Path("/Users/reeceway/Desktop/vericall voiceprints/commercial_safe_cleanroom/data/manifests/institutional_release_20260309_spoof/layer3_sv_manifest.csv"))
    parser.add_argument("--spoof-manifest", type=Path, default=Path("/Users/reeceway/Desktop/vericall voiceprints/commercial_safe_cleanroom/data/manifests/institutional_release_20260309_spoof/layer2_spoof_manifest.csv"))
    parser.add_argument("--speaker-model", type=Path, default=Path("/Users/reeceway/Desktop/vericall voiceprints/vericall/ios/VeriCall/Models/VoiceEmbedder.mlpackage"))
    parser.add_argument("--spoof-model", type=Path, default=Path("/Users/reeceway/Desktop/vericall voiceprints/vericall/ios/VeriCall/Models/VeriCallSpoofDetector.mlpackage"))
    parser.add_argument("--max-speakers", type=int, default=25)
    parser.add_argument("--max-spoof-per-class", type=int, default=50)
    args = parser.parse_args()

    print(f"speaker_manifest={args.speaker_manifest}")
    print(f"spoof_manifest={args.spoof_manifest}")
    print(f"speaker_model={args.speaker_model}")
    print(f"spoof_model={args.spoof_model}")

    speaker_metrics = run_speaker_benchmark(args.speaker_manifest, args.speaker_model, args.max_speakers)
    print("SPEAKER_BENCH")
    for key, value in speaker_metrics.items():
        print(f"{key}={value:.4f}")

    spoof_metrics = run_spoof_benchmark(args.spoof_manifest, args.spoof_model, args.max_spoof_per_class)
    print("SPOOF_BENCH")
    for key, value in spoof_metrics.items():
        print(f"{key}={value:.4f}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: ios/scripts/benchmark_webrtc_models.swift
#!/usr/bin/env swift

import Foundation
import CoreML
import AVFoundation
import Accelerate

struct AudioConfiguration {
    static let analysisWindowSamples = 48_000
    static let speakerThreshold: Float = 0.90
    static let spoofThreshold: Float = 0.665
}

struct CSVRow {
    let values: [String: String]
    subscript(_ key: String) -> String { values[key, default: ""] }
}

struct SpeakerExample {
    let speakerID: String
    let audioPath: String
}

struct SpeakerPairResult {
    let similarity: Float
    let predictedMatch: Bool
    let expectedMatch: Bool
}

struct SpoofExample {
    let label: String
    let audioPath: String
    let speakerID: String
}

struct SpoofResultRow {
    let cloneProbability: Float
    let predictedHuman: Bool
    let expectedHuman: Bool
}

enum BenchError: Error {
    case missingFeature(String)
    case invalidModelOutput(String)
}

final class LocalAudioLoader {
    func loadMono16kWindow(path: String, targetSamples: Int = AudioConfiguration.analysisWindowSamples) throws -> [Float] {
        let url = URL(fileURLWithPath: path)
        let file = try AVAudioFile(forReading: url)
        let sourceFormat = file.processingFormat
        guard let sourceBuffer = AVAudioPCMBuffer(
            pcmFormat: sourceFormat,
            frameCapacity: AVAudioFrameCount(file.length)
        ) else {
            throw NSError(domain: "Bench", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to create source buffer"])
        }
        try file.read(into: sourceBuffer)

        let targetFormat = AVAudioFormat(
            commonFormat: .pcmFormatFloat32,
            sampleRate: 16_000,
            channels: 1,
            interleaved: false
        )!

        let mono16kBuffer: AVAudioPCMBuffer
        if sourceFormat.sampleRate == targetFormat.sampleRate,
           sourceFormat.channelCount == targetFormat.channelCount,
           sourceFormat.commonFormat == .pcmFormatFloat32 {
            mono16kBuffer = sourceBuffer
        } else {
            guard let converter = AVAudioConverter(from: sourceFormat, to: targetFormat) else {
                throw NSError(domain: "Bench", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to create AVAudioConverter"])
            }
            let outputCapacity = AVAudioFrameCount(Double(sourceBuffer.frameLength) * targetFormat.sampleRate / sourceFormat.sampleRate) + 8
            guard let outputBuffer = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: outputCapacity) else {
                throw NSError(domain: "Bench", code: 3, userInfo: [NSLocalizedDescriptionKey: "Failed to create output buffer"])
            }
            var consumed = false
            let inputBlock: AVAudioConverterInputBlock = { _, outStatus in
                if consumed {
                    outStatus.pointee = .endOfStream
                    return nil
                }
                consumed = true
                outStatus.pointee = .haveData
                return sourceBuffer
            }
            var error: NSError?
            converter.convert(to: outputBuffer, error: &error, withInputFrom: inputBlock)
            if let error { throw error }
            mono16kBuffer = outputBuffer
        }

        let frameLength = Int(mono16kBuffer.frameLength)
        let channelData = mono16kBuffer.floatChannelData![0]
        var samples = Array(UnsafeBufferPointer(start: channelData, count: frameLength))
        if samples.count > targetSamples {
            samples = Array(samples.suffix(targetSamples))
        } else if samples.count < targetSamples {
            samples += Array(repeating: 0, count: targetSamples - samples.count)
        }
        return samples
    }
}

final class LocalSpeakerBench {
    private let model: MLModel
    private let sampleRate: Float = 16_000
    private let frameLength: Int = 400
    private let frameStep: Int = 160
    private let numMelBands: Int = 80
    private let numFrames: Int = 300

    init(modelURL: URL) throws {
        let compiled = try MLModel.compileModel(at: modelURL)
        let config = MLModelConfiguration()
        config.computeUnits = .cpuOnly
        self.model = try MLModel(contentsOf: compiled, configuration: config)
    }

    func embed(samples: [Float]) throws -> [Float] {
        let fbank = try computeFbank(samples: samples)
        let normalized = sentenceMeanSubtract(fbank)
        let input = try MLMultiArray(shape: [1, NSNumber(value: numFrames), NSNumber(value: numMelBands)], dataType: .float32)
        let ptr = input.dataPointer.assumingMemoryBound(to: Float.self)
        normalized.withUnsafeBufferPointer { src in
            ptr.initialize(from: src.baseAddress!, count: src.count)
        }
        let provider = try MLDictionaryFeatureProvider(dictionary: ["fbank_features": MLFeatureValue(multiArray: input)])
        let out = try model.prediction(from: provider)
        guard let arr = out.featureValue(for: "embedding")?.multiArrayValue else {
            throw BenchError.missingFeature("embedding")
        }
        let embPtr = arr.dataPointer.assumingMemoryBound(to: Float.self)
        return Array(UnsafeBufferPointer(start: embPtr, count: 192))
    }

    func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
        var dot: Float = 0
        var normA: Float = 0
        var normB: Float = 0
        vDSP_dotpr(a, 1, b, 1, &dot, vDSP_Length(a.count))
        vDSP_svesq(a, 1, &normA, vDSP_Length(a.count))
        vDSP_svesq(b, 1, &normB, vDSP_Length(b.count))
        let denom = sqrt(normA) * sqrt(normB)
        guard denom > 1e-8 else { return 0 }
        return max(0, min(1, dot / denom))
    }

    private func computeFbank(samples: [Float]) throws -> [Float] {
        var buf = samples
        if buf.count < AudioConfiguration.analysisWindowSamples {
            buf += Array(repeating: 0, count: AudioConfiguration.analysisWindowSamples - buf.count)
        } else if buf.count > AudioConfiguration.analysisWindowSamples {
            buf = Array(buf.prefix(AudioConfiguration.analysisWindowSamples))
        }

        var emphasized = [Float](repeating: 0, count: AudioConfiguration.analysisWindowSamples)
        emphasized[0] = buf[0]
        let alpha: Float = 0.97
        for i in 1..<AudioConfiguration.analysisWindowSamples {
            emphasized[i] = buf[i] - alpha * buf[i - 1]
        }

        let melFilters = Self.melFilterbank(numFilters: numMelBands, fftSize: frameLength, sampleRate: sampleRate)
        let fftSize = frameLength
        let halfFFT = fftSize / 2 + 1
        let log2n = vDSP_Length(log2(Float(fftSize)))
        guard let fftSetup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2)) else {
            throw NSError(domain: "Bench", code: 4, userInfo: [NSLocalizedDescriptionKey: "Failed to create FFT setup"])
        }
        defer { vDSP_destroy_fftsetup(fftSetup) }

        var result = [Float](repeating: 0, count: numFrames * numMelBands)
        var window = [Float](repeating: 0, count: fftSize)
        vDSP_hamm_window(&window, vDSP_Length(fftSize), 0)

        for frame in 0..<numFrames {
            let start = frame * frameStep
            let end = start + fftSize
            guard end <= emphasized.count else { break }
            var realPart = Array(emphasized[start..<end])
            vDSP_vmul(realPart, 1, window, 1, &realPart, 1, vDSP_Length(fftSize))
            var imagPart = [Float](repeating: 0, count: fftSize)
            realPart.withUnsafeMutableBufferPointer { rp in
                imagPart.withUnsafeMutableBufferPointer { ip in
                    var split = DSPSplitComplex(realp: rp.baseAddress!, imagp: ip.baseAddress!)
                    vDSP_fft_zip(fftSetup, &split, 1, log2n, FFTDirection(kFFTDirection_Forward))
                }
            }
            var power = [Float](repeating: 0, count: halfFFT)
            for i in 0..<halfFFT {
                power[i] = realPart[i] * realPart[i] + imagPart[i] * imagPart[i]
            }
            for m in 0..<numMelBands {
                var energy: Float = 0
                vDSP_dotpr(power, 1, melFilters[m], 1, &energy, vDSP_Length(halfFFT))
                result[frame * numMelBands + m] = log(max(energy, 1e-10))
            }
        }
        return result
    }

    private func sentenceMeanSubtract(_ fbank: [Float]) -> [Float] {
        var mean: Float = 0
        vDSP_meanv(fbank, 1, &mean, vDSP_Length(fbank.count))
        var out = fbank
        var neg = -mean
        vDSP_vsadd(out, 1, &neg, &out, 1, vDSP_Length(out.count))
        return out
    }

    private static var cachedMel: [[Float]] = []
    private static func melFilterbank(numFilters: Int, fftSize: Int, sampleRate: Float) -> [[Float]] {
        if !cachedMel.isEmpty { return cachedMel }
        let halfFFT = fftSize / 2 + 1
        func hzToMel(_ hz: Float) -> Float { 2595 * log10(1 + hz / 700) }
        func melToHz(_ mel: Float) -> Float { 700 * (pow(10, mel / 2595) - 1) }
        let melMin = hzToMel(0)
        let melMax = hzToMel(sampleRate / 2)
        let melPoints = (0...(numFilters + 1)).map { i -> Float in
            melToHz(melMin + Float(i) * (melMax - melMin) / Float(numFilters + 1))
        }
        let binPoints = melPoints.map { Int(($0 / sampleRate) * Float(fftSize) + 0.5) }
        var filters = [[Float]](repeating: [Float](repeating: 0, count: halfFFT), count: numFilters)
        for m in 0..<numFilters {
            let left = binPoints[m]
            let center = binPoints[m + 1]
            let right = binPoints[m + 2]
            if center > left {
                for k in left..<center where k < halfFFT {
                    filters[m][k] = Float(k - left) / Float(center - left)
                }
            }
            if right > center {
                for k in center..<right where k < halfFFT {
                    filters[m][k] = Float(right - k) / Float(right - center)
                }
            }
        }
        cachedMel = filters
        return filters
    }
}

final class LocalSpoofBench {
    private let model: MLModel

    init(modelURL: URL) throws {
        let compiled = try MLModel.compileModel(at: modelURL)
        let config = MLModelConfiguration()
        config.computeUnits = .cpuOnly
        self.model = try MLModel(contentsOf: compiled, configuration: config)
    }

    func predict(samples: [Float]) throws -> Float {
        let input = try MLMultiArray(shape: [1, NSNumber(value: AudioConfiguration.analysisWindowSamples)], dataType: .float32)
        let ptr = input.dataPointer.assumingMemoryBound(to: Float.self)
        for i in 0..<AudioConfiguration.analysisWindowSamples {
            ptr[i] = i < samples.count ? samples[i] : 0
        }
        let provider = try MLDictionaryFeatureProvider(dictionary: ["waveform": MLFeatureValue(multiArray: input)])
        let out = try model.prediction(from: provider)
        if let arr = out.featureValue(for: "clone_probability")?.multiArrayValue {
            return max(0, min(1, arr[0].floatValue))
        }
        if let val = out.featureValue(for: "clone_probability")?.doubleValue {
            return max(0, min(1, Float(val)))
        }
        throw BenchError.invalidModelOutput("clone_probability")
    }
}

func readCSVRows(at path: String) throws -> [CSVRow] {
    let text = try String(contentsOfFile: path, encoding: .utf8)
    let lines = text.split(whereSeparator: \.isNewline).map(String.init)
    guard let headerLine = lines.first else { return [] }
    let headers = headerLine.split(separator: ",").map(String.init)
    var rows: [CSVRow] = []
    for line in lines.dropFirst() {
        let cols = line.split(separator: ",", omittingEmptySubsequences: false).map(String.init)
        guard cols.count == headers.count else { continue }
        rows.append(CSVRow(values: Dictionary(uniqueKeysWithValues: zip(headers, cols))))
    }
    return rows
}

func runSpeakerBenchmark(manifestPath: String, modelPath: String, maxSpeakers: Int) throws {
    let rows = try readCSVRows(at: manifestPath)
    let filtered = rows.filter { $0["label"] == "real" && ($0["split"] == "val" || $0["split"] == "test") }
    var bySpeaker: [String: [SpeakerExample]] = [:]
    for row in filtered {
        bySpeaker[row["speaker_id"], default: []].append(SpeakerExample(speakerID: row["speaker_id"], audioPath: row["audio_path"]))
    }
    let speakers = bySpeaker.keys.sorted().filter { (bySpeaker[$0]?.count ?? 0) >= 2 }
    let selected = Array(speakers.prefix(maxSpeakers))
    let loader = LocalAudioLoader()
    let bench = try LocalSpeakerBench(modelURL: URL(fileURLWithPath: modelPath))

    var embeddings: [String: [Float]] = [:]
    var positives: [SpeakerPairResult] = []
    var negatives: [SpeakerPairResult] = []

    for speaker in selected {
        guard let examples = bySpeaker[speaker], examples.count >= 2 else { continue }
        let enroll = try loader.loadMono16kWindow(path: examples[0].audioPath)
        let probe = try loader.loadMono16kWindow(path: examples[1].audioPath)
        let enrollEmb = try bench.embed(samples: enroll)
        embeddings[speaker] = enrollEmb
        let probeEmb = try bench.embed(samples: probe)
        let sim = bench.cosineSimilarity(enrollEmb, probeEmb)
        positives.append(SpeakerPairResult(similarity: sim, predictedMatch: sim >= AudioConfiguration.speakerThreshold, expectedMatch: true))
    }

    for i in 0..<selected.count {
        let s1 = selected[i]
        let s2 = selected[(i + 1) % selected.count]
        guard s1 != s2,
              let enrollEmb = embeddings[s1],
              let examples = bySpeaker[s2], let probePath = examples.first?.audioPath else { continue }
        let probe = try loader.loadMono16kWindow(path: probePath)
        let probeEmb = try bench.embed(samples: probe)
        let sim = bench.cosineSimilarity(enrollEmb, probeEmb)
        negatives.append(SpeakerPairResult(similarity: sim, predictedMatch: sim >= AudioConfiguration.speakerThreshold, expectedMatch: false))
    }

    let all = positives + negatives
    let correct = all.filter { $0.predictedMatch == $0.expectedMatch }.count
    let posMean = positives.map(\.similarity).reduce(0, +) / Float(max(1, positives.count))
    let negMean = negatives.map(\.similarity).reduce(0, +) / Float(max(1, negatives.count))
    print("SPEAKER_BENCH")
    print("selected_speakers=\(selected.count)")
    print("positive_pairs=\(positives.count)")
    print("negative_pairs=\(negatives.count)")
    print(String(format: "speaker_accuracy=%.4f", Double(correct) / Double(max(1, all.count))))
    print(String(format: "positive_mean_similarity=%.4f", posMean))
    print(String(format: "negative_mean_similarity=%.4f", negMean))
    if let firstBad = all.first(where: { $0.predictedMatch != $0.expectedMatch }) {
        print(String(format: "speaker_first_error_similarity=%.4f predicted=%@ expected=%@", firstBad.similarity, String(firstBad.predictedMatch), String(firstBad.expectedMatch)))
    }
}

func runSpoofBenchmark(manifestPath: String, modelPath: String, maxPerClass: Int) throws {
    let rows = try readCSVRows(at: manifestPath)
    let filtered = rows.filter { $0["split"] == "val" || $0["split"] == "test" }
    let realRows = Array(filtered.filter { $0["label"] == "real" }.prefix(maxPerClass))
    let fakeRows = Array(filtered.filter { $0["label"] == "fake" }.prefix(maxPerClass))
    let loader = LocalAudioLoader()
    let bench = try LocalSpoofBench(modelURL: URL(fileURLWithPath: modelPath))
    var results: [SpoofResultRow] = []
    var realScores: [Float] = []
    var fakeScores: [Float] = []

    for row in realRows + fakeRows {
        let samples = try loader.loadMono16kWindow(path: row["audio_path"])
        let rms = sqrt(samples.reduce(0) { $0 + $1 * $1 } / Float(max(1, samples.count)))
        let gain = min(10.0 as Float, 0.03 / max(rms, 1e-6))
        let prepared = samples.map { max(-1, min(1, $0 * gain)) }
        let p = try bench.predict(samples: prepared)
        let expectedHuman = row["label"] == "real"
        let predictedHuman = p < AudioConfiguration.spoofThreshold
        results.append(SpoofResultRow(cloneProbability: p, predictedHuman: predictedHuman, expectedHuman: expectedHuman))
        if expectedHuman { realScores.append(p) } else { fakeScores.append(p) }
    }

    let correct = results.filter { $0.predictedHuman == $0.expectedHuman }.count
    let realMean = realScores.reduce(0, +) / Float(max(1, realScores.count))
    let fakeMean = fakeScores.reduce(0, +) / Float(max(1, fakeScores.count))
    print("SPOOF_BENCH")
    print("real_rows=\(realRows.count)")
    print("fake_rows=\(fakeRows.count)")
    print(String(format: "spoof_accuracy=%.4f", Double(correct) / Double(max(1, results.count))))
    print(String(format: "real_mean_clone_probability=%.4f", realMean))
    print(String(format: "fake_mean_clone_probability=%.4f", fakeMean))
    if let firstBad = results.first(where: { $0.predictedHuman != $0.expectedHuman }) {
        print(String(format: "spoof_first_error_clone_probability=%.4f predictedHuman=%@ expectedHuman=%@", firstBad.cloneProbability, String(firstBad.predictedHuman), String(firstBad.expectedHuman)))
    }
}

let speakerManifest = "/Users/reeceway/Desktop/vericall voiceprints/commercial_safe_cleanroom/data/manifests/institutional_release_20260309_spoof/layer3_sv_manifest.csv"
let spoofManifest = "/Users/reeceway/Desktop/vericall voiceprints/commercial_safe_cleanroom/data/manifests/institutional_release_20260309_spoof/layer2_spoof_manifest.csv"
let speakerModel = "/Users/reeceway/Desktop/vericall voiceprints/vericall/ios/VeriCall/Models/VoiceEmbedder.mlpackage"
let spoofModel = "/Users/reeceway/Desktop/vericall voiceprints/vericall/ios/VeriCall/Models/VeriCallSpoofDetector.mlpackage"

let speakerCount = CommandLine.arguments.count > 1 ? (Int(CommandLine.arguments[1]) ?? 20) : 20
let spoofPerClass = CommandLine.arguments.count > 2 ? (Int(CommandLine.arguments[2]) ?? 50) : 50

do {
    print("starting speaker benchmark speakers=\(speakerCount)")
    fflush(stdout)
    try runSpeakerBenchmark(manifestPath: speakerManifest, modelPath: speakerModel, maxSpeakers: speakerCount)
    print("starting spoof benchmark per_class=\(spoofPerClass)")
    fflush(stdout)
    try runSpoofBenchmark(manifestPath: spoofManifest, modelPath: spoofModel, maxPerClass: spoofPerClass)
} catch {
    fputs("benchmark failed: \(error)\n", stderr)
    exit(1)
}
